src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,460 @@
+    ]:
+        """
+        Returns a mapping of the most specific notification setting providers for the given recipients and scopes.
+        Note that this includes default settings for any notification types that are not set.
+        """
+
+        settings = self.get_all_setting_providers()
+        project_settings = self._filter_providers(
+            settings=settings, scope_type=NotificationScopeEnum.PROJECT.value, **kwargs
+        )
+        org_settings = self._filter_providers(
+            settings=settings, scope_type=NotificationScopeEnum.ORGANIZATION.value, **kwargs
+        )
+        user_settings = self._filter_providers(
+            settings=settings, scope_type=NotificationScopeEnum.USER.value, **kwargs
+        )
+        scoped_settings = [project_settings, org_settings, user_settings]
+
+        defaults = get_provider_defaults()
+        layered_setting_providers: MutableMapping[
+            Recipient,
+            MutableMapping[
+                Scope,
+                MutableMapping[
+                    NotificationSettingEnum,
+                    MutableMapping[ExternalProviderEnum, NotificationSettingsOptionEnum],
+                ],
+            ],
+        ] = defaultdict(lambda: defaultdict(lambda: defaultdict(dict)))
+        for recipient in self.recipients:
+            (scope_type, scope_items) = self._get_scope_for_layering(recipient)
+            for item in scope_items:
+                scope = (scope_type, item)
+                for type in NotificationSettingEnum:
+                    for provider in ExternalProviderEnum:
+                        most_specific_setting = None
+                        for setting in scoped_settings:
+                            user_settings = self._filter_providers(
+                                settings=setting, user_id=recipient.id, type=type.value
+                            )
+                            team_settings = self._filter_providers(
+                                settings=setting, team_id=recipient.id, type=type.value
+                            )
+                            s = user_settings + team_settings
+                            if len(s) > 0:
+                                most_specific_setting = NotificationSettingsOptionEnum(s[0].value)
+                                break
+
+                        if most_specific_setting is None:
+                            most_specific_setting = (
+                                NotificationSettingsOptionEnum.ALWAYS
+                                if provider in defaults
+                                else NotificationSettingsOptionEnum.NEVER
+                            )
+
+                        layered_setting_providers[recipient][scope][type].update(
+                            {provider: most_specific_setting}
+                        )
+
+        return layered_setting_providers
+
+    def get_notification_recipients(
+        self,
+        type: NotificationSettingEnum,
```

Comment:
@snigdhas we also have `self.type`, what if they are different?

Reply:
We can raise an error. It'll return 0 recipients otherwise, but calling it with the right params should be the caller's responsibility.

Reply:
@snigdhas makes sense but maybe we should throw an error if self.type is not none and it's different than the type passed it because it won't work